modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/js/components/FileEntryPicker.tsx

```diff
@@ -0,0 +1,321 @@
+import PropTypes from 'prop-types';
+import React, {ChangeEvent, useEffect, useRef, useState} from 'react';
+
+type TFeedback = {displayType: DisplayType; message: string};
+
+const EMPTY_FEEDBACK: TFeedback = {
+	displayType: 'info',
+	message: '',
+};
+
+const disableFormSubmitButton = (
+	disabled: boolean,
+	portletNamespace: string
+) => {
+	const submitButton = document.getElementById(
+		portletNamespace + 'editClientExtensionEntrySubmitButton'
+	) as HTMLButtonElement;
+
+	if (submitButton) {
+		submitButton.disabled = disabled;
+	}
+};
+
+const FileEntryPicker = ({
+	externalReferenceCode,
+	frontendTokenDefinition: currentFrontendTokenDefinition,
+	learnResources,
+	portletNamespace,
+}) => {
+	const [feedback, setFeedback] = useState(EMPTY_FEEDBACK);
+	const [alertRole, setAlertRole] = useState<'alert' | null>(null);
+	const [isValidatingJSON, setIsValidatingJSON] = useState(false);
+	const [frontendTokenDefinition, setFrontendTokenDefinition] = useState(
+		currentFrontendTokenDefinition
+	);
+
+	const fileInputRef = useRef<HTMLInputElement>();
+	const selectFileButtonRef = useRef<HTMLButtonElement>();
+
+	const clearSelection = () => {
+		setFeedback(EMPTY_FEEDBACK);
+
+		setFrontendTokenDefinition('');
+
+		disableFormSubmitButton(false, portletNamespace);
+
+		selectFileButtonRef.current?.focus();
+	};
+
+	const validateFrontendTokenDefinition = async (
+		body: any,
+		endpoint: string
+	) => {
+		try {
+			const response = await fetch(
+				'/o/com-liferay-frontend-token-definition-impl' + endpoint,
+				{
+					body,
+					method: 'POST',
+				}
+			);
+
+			const data = await response.json();
+
+			let displayType: DisplayType = 'success';
+
+			if (!response.ok) {
+				displayType = 'danger';
+
+				disableFormSubmitButton(true, portletNamespace);
+			}
+			else {
+				disableFormSubmitButton(false, portletNamespace);
+			}
+
+			setFeedback({
+				displayType,
+				message: data.message,
+			});
+		}
+		catch (error) {
+			clearSelection();
+
+			setFeedback({
+				displayType: 'danger',
+				message: Liferay.Language.get(
+					'your-upload-failed-to-complete-please-try-again-or-contact-support-if-the-error-persists'
+				),
+			});
+		}
+	};
+
+	const handleFileInputChange = async ({
+		target,
+	}: ChangeEvent<HTMLInputElement>) => {
+		const filePath = target.value;
+
+		const fileName = filePath.replace(/^.*[\\]/, '');
+
+		if (!fileName.endsWith('.json')) {
+			setFeedback({
+				displayType: 'danger',
+				message: Liferay.Language.get(
+					'the-format-is-not-valid-please-upload-a-valid-frontend-token-definition-json-file'
+				),
+			});
+
+			disableFormSubmitButton(true, portletNamespace);
+
+			return;
+		}
+
+		setIsValidatingJSON(true);
+
+		if (!alertRole) {
+			setAlertRole('alert');
+		}
+
+		setFeedback({
+			displayType: 'info',
+			message: Liferay.Language.get(
+				'the-frontend-token-definition-json-file-is-being-uploaded-and-validated'
+			),
+		});
+
+		if (target.files === null) {
+			return;
```

Comment:
If this ever happens it will leave us in a weird state where the feedback says we are still validating the json.